pkg/querier/queryrange/split_by_range.go

```diff
@@ -0,0 +1,120 @@
+package queryrange
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/weaveworks/common/httpgrpc"
+
+	"github.com/grafana/loki/pkg/loghttp"
+	"github.com/grafana/loki/pkg/logql"
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase"
+	"github.com/grafana/loki/pkg/tenant"
+	util_log "github.com/grafana/loki/pkg/util/log"
+	"github.com/grafana/loki/pkg/util/marshal"
+	"github.com/grafana/loki/pkg/util/validation"
+)
+
+type splitByRange struct {
+	logger log.Logger
+	next   queryrangebase.Handler
+	limits Limits
+
+	ng *logql.DownstreamEngine
+}
+
+// SplitByRangeMiddleware creates a new Middleware that splits log requests by the range interval.
+func SplitByRangeMiddleware(logger log.Logger, limits Limits, metrics *logql.ShardingMetrics) queryrangebase.Middleware {
+	return queryrangebase.MiddlewareFunc(func(next queryrangebase.Handler) queryrangebase.Handler {
+		return &splitByRange{
+			logger: log.With(logger, "middleware", "InstantQuery.splitByRangeVector"),
+			next:   next,
+			limits: limits,
+			ng:     logql.NewDownstreamEngine(logql.EngineOpts{}, DownstreamHandler{next}, metrics, limits, logger),
+		}
+	})
+}
+
+func (s *splitByRange) Do(ctx context.Context, request queryrangebase.Request) (queryrangebase.Response, error) {
+	logger := util_log.WithContext(ctx, s.logger)
+
+	tenants, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+	}
+
+	interval := validation.SmallestPositiveNonZeroDurationPerTenant(tenants, s.limits.QuerySplitDuration)
+	// if no interval configured, continue to the next middleware
+	if interval == 0 {
+		return s.next.Do(ctx, request)
+	}
+
+	mapper, err := logql.NewRangeVectorMapper(interval)
+	if err != nil {
+		return nil, err
+	}
+
+	noop, parsed, err := mapper.Parse(request.GetQuery())
+	if err != nil {
+		level.Warn(logger).Log("msg", "failed mapping AST", "err", err.Error(), "query", request.GetQuery())
+		return nil, err
+	}
+	level.Debug(logger).Log("msg", "mapped instant query", "interval", interval.String(), "noop", noop, "original", request.GetQuery(), "mapped", parsed.String())
+
+	if noop {
+		// the query cannot be split, so continue
+		return s.next.Do(ctx, request)
+	}
+
+	params, err := paramsFromRequest(request)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, ok := request.(*LokiInstantRequest); !ok {
+		return nil, fmt.Errorf("expected *LokiInstantRequest")
+	}
+
+	query := s.ng.Query(params, parsed)
+
+	res, err := query.Exec(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	value, err := marshal.NewResultValue(res.Data)
+	if err != nil {
+		return nil, err
+	}
+
+	switch res.Data.Type() {
```

Comment:
`NewQueryShardMiddleware` has the same dance. What do you think about extracting this logic? Maybe it's not useful when just two function use it.

Reply:
I'll see if I can extract functions in a meaningful way.

Reply:
After looking at both middlewares, I think it does not make a lot of sense to try to abstract logic used by both. I do think that the middleware structure will change in the future in favour of a more conventional query planner anyway.